### PR TITLE
[style edition 2024] Combine all delimited exprs as last argument

### LINKED ISSUE #149: Delimited expressions should be allowed as last argument
(this was originally a [comment in #61](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-573370023))

I'm trying to dig up why the rustfmt `overflow_delimited_expr` shouldn't be enabled by default. And I found it derives from the following statement in the [guide](https://github.com/rust-dev-tools/fmt-rfcs/blob/master/guide/expressions.md#combinable-expressions)

> Only where the multi-line sub-expression is a closure with an explicit block, this combining behaviour may be used where there are other arguments, as long as all the arguments and the first line of the closure fit on the first line, the closure is the last argument, and there is only one closure argument

### History

I see in [this comment](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issue-206470936) @joshtriplett says

> This should never apply if the outer construct contains more than just the single expression
> ```rust
> // don't do this
> let thing = Struct(first, OtherStruct(
>     second,
>     third,
> ));
> 
> // don't do this either
> foo(thing, bar(
>     param1,
>     param2,
>     param3,
> ));
> ```

In [the next comment](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-278613680) they suggest a possible exception for format macros

> As a possible exception to the requirement that the outer construct must have only one argument, you can also apply this to an invocation of `print!`, `println!`, or `format!`, where the first argument consists of a format string and the second argument consists of a single expression:
> 
> ```rust
> println!("Thing: {}", generate_thing(
>     long_argument_one,
>     long_argument_two,
>     long_argument_three,
> ));
> ```

_AFAIK this did not make it into the guide_

Then @nrc suggests an exception for closures as the last argument in [this comment](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-300055628) and @joshtriplett [agrees](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-302234686)

> Re closures, I would like to restrict this to block closures, and in this case we allow any number of parameters as long as everything up to the { of the closure fits on one line (including inside the 'short' heuristic for function args), e.g.,:
> ```rust
> foo(1 + 2, x, some_variable.bar(), |x| {
>     closure_body();
>     ret_value
> })
> ```

_This did make it into the guide_

Then there's a lot of discussion about what control flow expressions should be allowed.

Then there's discussion about arrays where @kennytm [brings up arrays](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-311551291) and @joshtriplett [agrees](https://github.com/rust-dev-tools/fmt-rfcs/issues/61#issuecomment-311562570) that it should be treated similarly.

However, there was never any discussion about allowing arrays, match, and other delimited expressions as the last argument.

### Getting to the point

I don't see any reason why other multi-line delimited expressions shouldn't be allowed by default in the same context as closures. For instance, right now, rustfmt will format a match inside a closure and a match without a closure differently:
```rust
    // with a closure
    let f = fn_call(123, |cond| match cond {
        A => 1,
        B => 2,
    });
    // without a closure, not consistent
    let f = fn_call(
        123,
        match cond {
            A => 1,
            B => 2,
        },
    );
    // works fine with normal blocks though
    let f = fn_call(123, {
        let a = 1;
        let b = 2;
        a + b
    });
    // or without any other arguments
    let f = fn_call(match cond {
        A => 1,
        B => 2,
    });
```

That doesn't make _any_ sense to me, and looks super ugly. Also, consider an `append` macro, used to push a list of elements to a Vec without copying or cloning:
```rust
     append!(v, [
        352732685652347985,
        348974392769832796759287,
        4738973482759382759832,
        178957349857932759834729857,
    ]);
```

Awesome for cleaning up dozens of lines of `v.push(...)` calls into an easily readable format. But rustfmt just kills me here (without enabling the `overflow_delimited_expr` option):
```rust
     append!(
        v,
        [
            352732685652347985,
            348974392769832796759287,
            4738973482759382759832,
            178957349857932759834729857,
        ]
    );
```

So my question is, why does this "last argument exception" not apply to match, arrays, `vec![]`, etc? Seems inconsistent, yet alone ugly.

### PR DESCRIPTION
**Pull request by pitaj**
Closes rust-lang/style-team#149

If this is merged, the rustfmt option `overflow_delimited_expr` should be enabled by default in style edition 2024.

[Rendered](https://github.com/pitaj/rust/blob/style-delimited-expressions/src/doc/style-guide/src/expressions.md#combinable-expressions)

r? joshtriplett

### CODE REVIEW

File: src/doc/style-guide/src/expressions.md [hunk 1]
```diff
@@ -831,20 +831,56 @@ let arr = [combinable(
...
+let x = Thing(an_expr, another_expr, match cond {
+    A => 1,
+    B => 2,
+});
+
+let x = format!("Stuff: {}", [
+    an_expr,
+    another_expr,
+]);
 ```
```
Comment (joshtriplett) [inline src/doc/style-guide/src/expressions.md:844]:
``````suggestion

let x = func(an_expr, another_expr, SomeStruct {
    field: this_is_long,
    another_field: 123,
});
```
``````

File: src/doc/style-guide/src/expressions.md [hunk 2]
```diff
@@ -831,20 +831,56 @@ let arr = [combinable(
...
 
 Apply this behavior recursively.
 
-For a function with multiple arguments, if the last argument is a multi-line
-closure with an explicit block, there are no other closure arguments, and all
-the arguments and the first line of the closure fit on the first line, use the
-same combining behavior:
+If the last argument is a multi-line closure with an explicit block,
+only apply the combining behavior if there are no other closure arguments.
```
Comment (joshtriplett) [inline src/doc/style-guide/src/expressions.md:849]: Do we want to keep this property? Here's an example that this wouldn't allow, but that seems *relatively* reasonable:

```rust
func(arg, || expr, |closure_arg| {
    long_closure_here();
    more_of_the_long_closure();
});
```

We might want to prohibit this formatting, but let's consider this case specifically before assuming we want to preserve this restriction.

**Comment by joshtriplett** [comment]
@rfcbot merge

File: src/doc/style-guide/src/editions.md [hunk 3]
```diff
@@ -36,6 +36,9 @@ For a full history of changes in the Rust 2024 style edition, see the git
 history of the style guide. Notable changes in the Rust 2024 style edition
 include:
 
+- As the last argument of a function call, delimited expressions are generally
```
Comment (pitaj) [inline src/doc/style-guide/src/editions.md:39]: Should this link to the PR?

**Comment by calebcartwright** [comment]
@rfcbot concern require all checkboxes

(note I don't have any specific concerns, haven't even had a chance to read yet, it's just that given the small team size we require all 4 approvals)

**Comment by pitaj** [comment]
I recently learned of [this issue](https://github.com/rust-lang/rustfmt/issues/3370#issuecomment-1288226646). 

It brings up that there are currently some inconsistencies with how the `overflow_delimited_expr` option is implemented.

Changing their example from arrays of arrays to arrays of tuples completed changes the formatting behavior, which I don't think should be the case.

There are also cases where applying this option collapses in the wrong position, like the following:

```
// Expected
vbuf.write(i * 3, Bar(
    Foo { x: 1, y: 2, z: 3 },
    Foo { x: 1, y: 2, z: 3 },
    Foo { x: 1, y: 2, z: 3 },
));

// Actual
vbuf.write(
    i * 3,
    Bar(Foo { x: 1, y: 2, z: 3 }, Foo { x: 1, y: 2, z: 3 }, Foo {
        x: 1,
        y: 2,
        z: 3,
    }),
);
```

I'm going to think about how to specify the desired behavior for these cases. My first thought is something like this:

> Applying this role recursively should use the level of recursion that results in the minimal number of lines. When two levels of recursion have equal numbers of lines, the shallower level should be used.

**Comment by pitaj** [comment]
That might be a good idea to include regardless, but doesn't cover cases like the following:

```rust
vbuf.write(i + 3, &[
    [h - 0.1, v],
    [h - 0.1, v],
    [h - 0.1, v],
    [h - 0.1, v],
]);

vbuf.write(i + 3, (
    Foo { x, y },
    Bar { z, w },
    Thing { a, b },
));
```

A better idea might be to apply the same rule that closures with explicit blocks get:

> If the last argument is a struct literal, only apply the combining behavior if there are no other struct literals.

> If the last argument is an array literal, only apply the combining behavior if there are no other array literals.

Etc.

**Comment by calebcartwright** [comment]
I'm generally in favor of doing this, though I'm still holding off on providing my :heavy_check_mark: as I feel there's some unanswered questions/additions (e.g. https://github.com/rust-lang/rust/pull/114764#discussion_r1294838907) and it's unclear to me whether those have been implicitly addressed or if perhaps they were accidentally overlooked?

**Comment by pitaj** [comment]
@calebcartwright good point. I suppose my most recent change kinda implicitly answered Josh's concern, by expanding the rule across all kinds of expressions.

To directly answer it, I agree with him that the rule has value and should stay. The reason I didn't respond to it before is that I thought it was for T-style discussion rather than aimed at me.

It does bring up one more minor question: Any closure can prevent an explicit-body closure from being combined. Should the same apply to tuple vs field structs/enums?

My preemptive answer is that this is undecidable without type information, so we shouldn't even try. Even if it were possible without, tuple struct/enum syntax is more similar to a function call than to the field struct/enum. For closures it makes sense, since a closure without an explicit body doesn't have a clear terminator, unlike functions calls and tuple structs/enums which terminate with a closing paren.

**Comment by joshtriplett** [comment]
@pitaj I don't think that last change is as clear-cut as it looks. On the one hand, if you have a few short struct initializers, the examples you give look compelling. On the other hand, with a couple of short struct initializers (1-2 fields) followed by a long initializer (with a lot more fields), I think it makes sense to break the long initializer across lines and keep the short initializers in the same line:

```
func(Small { a: b }, AlsoSmall { x }, Long {
    field,
    field2: value,
    field3: another_value,
    field4: long_expression(),
    field5,
});
```

So I would propose that we don't have this restriction. The last commit was also added after the FCP. Could you please drop the last commit?

There's a long history of trying to find some general rule that avoids splitting up matrices or similar things, and tries to visually balance or align similar things; the problem is we've never found a sufficiently general rule that works in the general case. I don't think this specific rule works in the general case.

That said, separate from this, I'd also love to see us having complexity-based rules for when to break lines; by way of example, I *do* think that `func((a, b, c), (d, e, f))` is something we should break across lines, because it's visually hard to parse (with commas at two levels on the same line). That's not a chaining-specific rule, though.

**Comment by pitaj** [comment]
I'll revert the commit next chance I get, sorry about that.

The purpose of the change is not to improve formatting, but to prevent formatting regressions. These cases are currently formatted well, it would be unfortunate to make them significantly worse with this change.

**Comment by calebcartwright** [comment]
@rfcbot resolved require all checkboxes

**Comment by pitaj** [comment]
I'm on vacation so it may be a few days before I can squash the commits here.

I plan on following up with another PR for the "same expression kind" rule.

**Comment by pitaj** [comment]
FYI, this has been squashed and should be ready to merge.

**Comment by ytmimi** [comment]
I saw [this comment on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/346005-t-style/topic/PR.20114764/near/416831573), which linked to this PR. I just want to make sure I'm not missing anything. Would merging this PR require us to make changes to rustfmt? I'm also noticing that [overflow_delimited_expr](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#overflow_delimited_expr) is currently an unstable configuration option. Does it make sense to change the default give that the option is unstable? @calebcartwright is the plan to stabilize the option in rustfmt?

**Comment by pitaj** [comment]
It does require rustfmt changes.

I don't think this necessitates stabilizing the feature, but I don't see why you wouldn't given all the work would need to be done for the edition anyways.

**Comment by pitaj** [comment]
To be clear, I think it absolutely makes sense to stabilize the option as well, I was just saying that technically it's not absolutely required.

**Comment by ehuss** [comment]
Is there anything tracking for updating rustfmt to include this change?

**Comment by ytmimi** [comment]
Not to my knowledge, no. I don't think we've added anything on the rustfmt side to specifically to track the work for this change

**Comment by calebcartwright** [comment]
responding to the various comments about rustfmt stabilization:

in general, rustfmt having a configuration option that has different default values depending on the specified style edition value is not only permissible but essentially expected given the nature of the style edition as the mechanism to functionally evolve default styles over time

whether or not a configuration option meets all the criteria for stabilization as required by rustfmt's stabilization process strikes me as orthogonal; rustfmt allowing users on stable to specify any supported value is different from rustfmt's default value.

@ytmimi IIRC we already discussed this, but if not and if you've still got concerns then we can review, and if necessary follow up with the style team. Determining what makes the final cut for a style edition is going to be a collaborative process between the style and rustfmt teams, and if there's implementation timeline challenges then we'll let style team know and adjust accordingly.

That being said, I do not personally anticipate any issues with this one
